app/src/main/java/rkr/simplekeyboard/inputmethod/latin/utils/CapsModeUtils.java

```diff
@@ -136,15 +138,18 @@ public static int getCapsMode(final CharSequence cs, final int reqModes,
             return (TextUtils.CAP_MODE_CHARACTERS | TextUtils.CAP_MODE_WORDS
                     | TextUtils.CAP_MODE_SENTENCES) & reqModes;
         }
-        if (i == j) {
-            // If we don't have whitespace before index i, it means neither MODE_WORDS
+        if (newCapIndex == i) {
+            if (spacingAndPunctuations.isWordSeparator(prevChar)) {
```

Comment:
As per comment, if the character before is not a whitespace, we don't need to check if capitalization is needed. In `abcd,ef` there is no point for capitalization of letter `e`.

Reply:
The issue is that there is an inconsistency between auto-capitalization and recapitalization. For example in a title case field, if I type "s&p index" without doing any manual capitalization, it will result in "S&p index", but if I select the text and cycle through the recapitalization options, it gives me "S&P Index", "S&P INDEX", and "s&p index". `capitalizeEachWord` in `StringUtils` checks sorted separators, which comes from `SpacingAndPunctuations` and does the same check as `isWordSeparator`. If you don't want the letter after a separator capitalized, we should remove that check from `capitalizeEachWord`. I don't care much either way, but I think they should at least function the same.

Reply:
Ok, this explains a bit more. If you'll leave only this part then it can be merged (after I'll test it). GBoard also works differently.

Reply:
From my local testing this has no effect on _S&P Index_ capitalization. What exactly should this fix?

Reply:
In a title case field (I've been testing with the Company field in the default Contacts app), I type "S&" (never pressing the shift key) and the keyboard now shows all letters as capitalized (assuming auto-capitalization is enabled). Before this change all of the letters on the keyboard were lowercase. I checked and Gboard behaves the same (ironically, when doing recapitalization Gboard uses "S&p Index", which is the opposite inconsistency issue that I'm trying to address).

Reply:
With this flag https://developer.android.com/reference/android/text/InputType#TYPE_TEXT_FLAG_CAP_WORDS and your change the result is `S&P Index` which is consistent with GBoard.

However, for other separators like `()[]{}` it's inconsistent:
- `S(p Index`
- `S)P Index`

This is caused by `Step 2 : Skip (ignore at the end of input) any opening punctuation.` above. I can't tell what further side effects removal of this logic would cause.

Reply:
Good catch. I added a fix for that. This should only check the word separator for the last character.

We don't want to remove step 2. That allows capitalizing things like"Hello world! (**M**ore text.)" in sentence case or "Hello '**W**orld'" in title case.